File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/ChannelMarkerManager.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package edu.uci.ics.amber.engine.architecture.worker

import edu.uci.ics.amber.engine.architecture.messaginglayer.InputGateway
import edu.uci.ics.amber.engine.architecture.rpc.controlcommands.ChannelMarkerPayload
import edu.uci.ics.amber.engine.architecture.rpc.controlcommands.ChannelMarkerType.{
  NO_ALIGNMENT,
  REQUIRE_ALIGNMENT
}
import edu.uci.ics.amber.engine.common.{AmberLogging, CheckpointState}
import edu.uci.ics.amber.core.virtualidentity.{
  ActorVirtualIdentity,
  ChannelIdentity,
  ChannelMarkerIdentity
}

import scala.collection.mutable

class ChannelMarkerManager(val actorId: ActorVirtualIdentity, inputGateway: InputGateway)
    extends AmberLogging {

  private val markerReceived =
    new mutable.HashMap[ChannelMarkerIdentity, Set[ChannelIdentity]]()

  val checkpoints = new mutable.HashMap[ChannelMarkerIdentity, CheckpointState]()

  /**
    * Determines if an epoch marker is fully received from all relevant senders within its scope.
    * This method checks if the epoch marker, based on its type, has been received from all necessary channels.
    * For markers requiring alignment, it verifies receipt from all senders in the scope. For non-aligned markers,
    * it checks if it's the first received marker. Post verification, it cleans up the markers.
    *
    * @return Boolean indicating if the epoch marker is completely received from all senders
    *         within the scope. Returns true if the marker is aligned, otherwise false.
    */
  def isMarkerAligned(
      from: ChannelIdentity,
      marker: ChannelMarkerPayload
  ): Boolean = {
    val markerId = marker.id
    if (!markerReceived.contains(markerId)) {
      markerReceived(markerId) = Set()
    }
    markerReceived.update(markerId, markerReceived(markerId) + from)
    // check if the epoch marker is completed
    val markerReceivedFromAllChannels =
      getChannelsWithinScope(marker).subsetOf(markerReceived(markerId))
    val epochMarkerCompleted = marker.markerType match {
      case REQUIRE_ALIGNMENT => markerReceivedFromAllChannels
      case NO_ALIGNMENT      => markerReceived(markerId).size == 1 // only the first marker triggers
      case _ =>
        throw new IllegalArgumentException(
          s"Unsupported marker type: ${marker.markerType}"
        )
    }
    if (markerReceivedFromAllChannels) {
      markerReceived.remove(markerId) // clean up if all markers are received
    }
    epochMarkerCompleted
  }

  private def getChannelsWithinScope(marker: ChannelMarkerPayload): Set[ChannelIdentity] = {
    val upstreams = marker.scope.filter(_.toWorkerId == actorId)
    inputGateway.getAllChannels
      .map(_.channelId)
      .filter { id =>
        upstreams.contains(id)
      }
      .toSet
  }

}

File: core/amber/src/main/python/core/architecture/managers/channel_marker_manager.py
from collections import defaultdict
from typing import Set, Dict

from core.architecture.packaging.input_manager import Channel
from proto.edu.uci.ics.amber.core import ActorVirtualIdentity, ChannelIdentity
from proto.edu.uci.ics.amber.engine.architecture.rpc import (
    ChannelMarkerPayload,
    ChannelMarkerType,
)


class ChannelMarkerManager:
    def __init__(self, actor_id: ActorVirtualIdentity, input_gateway):
        self.actor_id = actor_id
        self.input_gateway = input_gateway
        self.marker_received: Dict[str, Set[ChannelIdentity]] = defaultdict(set)

    def is_marker_aligned(
        self, from_channel: ChannelIdentity, marker: ChannelMarkerPayload
    ) -> bool:
        """
        Checks whether a channel marker has been received from all expected
        input channels, determining whether further processing can proceed.

        Args:
            from_channel (ChannelIdentity): The channel from which the marker
                was received.
            marker (ChannelMarkerPayload): The marker payload containing its
                type and scope.

        Returns:
            bool: True if the marker is considered aligned and processing can
                  continue, False otherwise.
        """
        marker_id = marker.id
        self.marker_received[marker_id].add(from_channel)

        marker_received_from_all_channels = self.get_channels_within_scope(
            marker
        ).issubset(self.marker_received[marker_id])

        if marker.marker_type == ChannelMarkerType.REQUIRE_ALIGNMENT:
            epoch_marker_completed = marker_received_from_all_channels
        elif marker.marker_type == ChannelMarkerType.NO_ALIGNMENT:
            epoch_marker_completed = (
                len(self.marker_received[marker_id]) == 1
            )  # Only the first marker triggers
        else:
            raise ValueError(f"Unsupported marker type: {marker.marker_type}")

        if marker_received_from_all_channels:
            del self.marker_received[marker_id]  # Clean up if all markers are received

        return epoch_marker_completed

    def get_channels_within_scope(
        self, marker: ChannelMarkerPayload
    ) -> Dict["ChannelIdentity", "Channel"].keys:
        upstreams = {
            channel_id
            for channel_id in marker.scope
            if channel_id.to_worker_id == self.actor_id
        }
        return self.input_gateway.get_all_channel_ids() & upstreams
